### ISSUE TITLE
SSLError : SSL: DECRYPTION_FAILED_OR_BAD_RECORD_MAC - CSS images not downloaded

### ISSUE THREAD

**Issue by mukuntharajaa** [lorenzodifuccia/safaribooks#163]
I am on master commit ( 32bbe893e62a9ba7053297eb496d262e14fd4b89 ). While trying to download **"9781492047834"**, some of the CSS are not downloaded with following error.

[-] Downloading book contents... (16 chapters)                                                                                      
   [#########################################################################################################################] 100%
[-] Downloading book CSSs... (4 files)                                                                                              
**[#] HTTPSConnectionPool(host='learning.oreilly.com', port=443): Max retries exceeded with url: /static/css/annotator.e3b0c44298fc.css (Caused by SSLError(SSLError(1, '[SSL: WRONG_VERSION_NUMBER] wrong version number (_ssl.c:2508)')))
[#] HTTPSConnectionPool(host='learning.oreilly.com', port=443): Max retries exceeded with url: /static/CACHE/css/output.68851547a55f.css (Caused by SSLError(SSLError(1, '[SSL: DECRYPTION_FAILED_OR_BAD_RECORD_MAC] decryption failed or bad record mac (_ssl.c:2508)')))
[#] Error trying to retrieve this CSS: /home/xxxxx/Downloads/git/safaribooks/Books/Monolith to Microservices (9781492047834)/OEBPS/Styles/Style02.css
    From: https://learning.oreilly.com/static/css/annotator.e3b0c44298fc.css**
[#] Error trying to retrieve this CSS: /home/xxxxx/Downloads/git/safaribooks/Books/Monolith to Microservices (9781492047834)/OEBPS/Styles/Style01.css
    From: https://learning.oreilly.com/static/CACHE/css/output.68851547a55f.css
Process Process-3:
Process Process-2:
Traceback (most recent call last):
Traceback (most recent call last):
  File "/usr/lib64/python3.7/multiprocessing/process.py", line 297, in _bootstrap
    self.run()
  File "/usr/lib64/python3.7/multiprocessing/process.py", line 297, in _bootstrap
    self.run()
  File "/usr/lib64/python3.7/multiprocessing/process.py", line 99, in run
    self._target(*self._args, **self._kwargs)
  File "safaribooks.py", line 829, in _thread_download_css
    s.write(response.content)
  File "/usr/lib64/python3.7/multiprocessing/process.py", line 99, in run
    self._target(*self._args, **self._kwargs)
  File "safaribooks.py", line 829, in _thread_download_css
    s.write(response.content)
AttributeError: 'int' object has no attribute 'content'
AttributeError: 'int' object has no attribute 'content'
[-] Downloading book images... (132 files)                                                                                          
    [#########################################################################################################################] 100%
[-] Creating EPUB file...                                                                                                           
[*] Done: /home/xxxxx/Downloads/git/safaribooks/Books/Monolith to Microservices (9781492047834)/9781492047834.epub                 

    If you like it, please * this project on GitHub to make it known:
        https://github.com/lorenzodifuccia/safaribooks
    e don't forget to renew your Safari Books Online subscription:
        https://learning.oreilly.com

[!] Bye!!

]$ python3 --version
Python 3.7.5

**Comment by mustafagonul** [comment]
Have similar problem:

[-] Downloading book CSSs... (4 files)                                                                                                                                                        
[#] HTTPSConnectionPool(host='learning.oreilly.com', port=443): Max retries exceeded with url: /static/CACHE/css/output.68851547a55f.css (Caused by SSLError(SSLError(1, '[SSL: DECRYPTION_FAILED_OR_BAD_RECORD_MAC] decryption failed or bad record mac (_ssl.c:2560)')))
Process Process-3:
[#] Error trying to retrieve this CSS: /home/mv/GIT/GITHUB/safaribooks/Books/Software Architecture in Practice Third Edition (9780132942799)/OEBPS/Styles/Style01.css                         
    From: https://learning.oreilly.com/static/CACHE/css/output.68851547a55f.css
Process Process-2:
Traceback (most recent call last):
Traceback (most recent call last):
  File "/usr/lib/python3.8/multiprocessing/process.py", line 313, in _bootstrap
    self.run()
  File "/usr/lib/python3.8/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "safaribooks.py", line 824, in _thread_download_css
    response = self.requests_provider(url)
  File "safaribooks.py", line 414, in requests_provider
    response = getattr(self.session, "post" if is_post else "get")(
  File "/usr/lib/python3.8/multiprocessing/process.py", line 313, in _bootstrap
    self.run()
  File "/usr/lib/python3.8/site-packages/requests/sessions.py", line 546, in get
    return self.request('GET', url, **kwargs)
  File "/usr/lib/python3.8/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib/python3.8/site-packages/requests/sessions.py", line 533, in request
    resp = self.send(prep, **send_kwargs)
  File "safaribooks.py", line 829, in _thread_download_css
    s.write(response.content)
  File "/usr/lib/python3.8/site-packages/requests/sessions.py", line 686, in send
    r.content
  File "/usr/lib/python3.8/site-packages/requests/models.py", line 828, in content
    self._content = b''.join(self.iter_content(CONTENT_CHUNK_SIZE)) or b''
AttributeError: 'int' object has no attribute 'content'
  File "/usr/lib/python3.8/site-packages/requests/models.py", line 750, in generate
    for chunk in self.raw.stream(chunk_size, decode_content=True):
  File "/usr/lib/python3.8/site-packages/urllib3/response.py", line 564, in stream
    data = self.read(amt=amt, decode_content=decode_content)
  File "/usr/lib/python3.8/site-packages/urllib3/response.py", line 507, in read
    data = self._fp.read(amt) if not fp_closed else b""
  File "/usr/lib/python3.8/http/client.py", line 454, in read
    n = self.readinto(b)
  File "/usr/lib/python3.8/http/client.py", line 498, in readinto
    n = self.fp.readinto(b)
  File "/usr/lib/python3.8/socket.py", line 669, in readinto
    return self._sock.recv_into(b)
  File "/usr/lib/python3.8/ssl.py", line 1241, in recv_into
    return self.read(nbytes, buffer)
  File "/usr/lib/python3.8/ssl.py", line 1099, in read
    return self._sslobj.read(len, buffer)
ssl.SSLError: [SSL: DECRYPTION_FAILED_OR_BAD_RECORD_MAC] decryption failed or bad record mac (_ssl.c:2560)
[-] Downloading book images... (212 files)                                                                                                                                                    
    [###################################################################################################################################################################################] 100%
[-] Creating EPUB file...                                                                                                                                                                     
[*] Done: /home/mv/GIT/GITHUB/safaribooks/Books/Software Architecture in Practice Third Edition (9780132942799)/9780132942799.epub                                                            

    If you like it, please * this project on GitHub to make it known:
        https://github.com/lorenzodifuccia/safaribooks
    e don't forget to renew your Safari Books Online subscription:
        https://learning.oreilly.com

[!] Bye!!

Python 3.8.0

**Comment by mukuntharajaa** [comment]
@lorenzodifuccia Is this not a bug ? If its not bug in the code and problem lies somewhere else, can you please point out where is it going wrong ?

**Comment by ktaild** [comment]
```
    def collect_css(selThis issue maybe related with the issue of requests module which safaribooks.py using:
[OpenSSL.SSL.Error: [('SSL routines', 'SSL3_GET_RECORD', 'decryption failed or bad record mac')] · Issue #1906 · psf/requests](https://github.com/psf/requests/issues/1906)

You probably workaround this issue by avoiding to use multi-threading.
e.g., modifying safaribooks.py as below:
        self.display.state_status.value = -1

        if "win" in sys.platform:
            # TODO
            for css_url in self.css:
                self._thread_download_css(css_url)

        else:
           # self._start_multiprocessing(self._thread_download_css, self.css)
            for css_url in self.css:
                self._thread_download_css(css_url)
```

```
    def collect_images(self):
        if self.display.book_ad_info == 2:
            self.display.info("Some of the book contents were already downloaded.\n"
                              "    If you want to be sure that all the images will be downloaded,\n"
                              "    please delete the output direcotry '" + self.BOOK_PATH +
                              "' and restart the program.")

        self.display.state_status.value = -1

        if "win" in sys.platform:
            # TODO
            for image_url in self.images:
                self._thread_download_images(image_url)

        else:
            #self._start_multiprocessing(self._thread_download_images, self.images)
            for image_url in self.images:
                self._thread_download_images(image_url)
```

**Comment by mukuntharajaa** [comment]
If someone is hitting this thread, following work around fixes the problem.

diff --git a/safaribooks.py b/safaribooks.py
index 95c13a3..8d86b49 100644
--- a/safaribooks.py
+++ b/safaribooks.py
@@ -16,7 +16,6 @@ from lxml import html, etree
 from multiprocessing import Process, Queue, Value
 from urllib.parse import urljoin, urlparse, parse_qs, quote_plus

-
 PATH = os.path.dirname(os.path.realpath(__file__))
 COOKIES_FILE = os.path.join(PATH, "cookies.json")

@@ -880,7 +879,9 @@ class SafariBooks:
                 self._thread_download_css(css_url)

         else:
-            self._start_multiprocessing(self._thread_download_css, self.css)
+            for css_url in self.css:
+                self._thread_download_css(css_url)
+            #self._start_multiprocessing(self._thread_download_css, self.css)

     def collect_images(self):
         if self.display.book_ad_info == 2:
@@ -897,7 +898,9 @@ class SafariBooks:
                 self._thread_download_images(image_url)

         else:
-            self._start_multiprocessing(self._thread_download_images, self.images)
+            #self._start_multiprocessing(self._thread_download_images, self.images)
+            for image_url in self.images:
+                self._thread_download_images(image_url)

     def create_content_opf(self):

**Comment by TheSnoozer** [comment]
Perhaps it's worth to point out that even with the suggested workaround the download still can fail:
```
[DATE] Downloading book CSSs... (4 files)
[DATE] HTTPSConnectionPool(host='maxcdn.bootstrapcdn.com', port=443): Max retries exceeded with url: /font-awesome/4.4.0/css/font-awesome.min.css (Caused by NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x034A2DA8>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))
[DATE] Error trying to retrieve this CSS: .......OEBPS\Styles\Style03.css
    From: https://maxcdn.bootstrapcdn.com/font-awesome/4.4.0/css/font-awesome.min.css
[DATE]  File ".......\safaribooks\safaribooks.py", line 1093, in <module>
    SafariBooks(args_parsed)
  File ".......\safaribooks\safaribooks.py", line 391, in __init__
    self.collect_css()
  File ".......\safaribooks\safaribooks.py", line 880, in collect_css
    self._thread_download_css(css_url)
  File ".......\safaribooks\safaribooks.py", line 832, in _thread_download_css
    s.write(response.content)

[DATE] Unhandled Exception: 'int' object has no attribute 'content' (type: AttributeError)
```

Note here content is equal to `int(0)` and hence does not have a content attribute.

**Comment by nchiarappa** [comment]
> If someone is hitting this thread, following work around fixes the problem.
> 
> ```
> diff --git a/safaribooks.py b/safaribooks.py
> index 95c13a3..8d86b49 100644
> --- a/safaribooks.py
> +++ b/safaribooks.py
> @@ -16,7 +16,6 @@ from lxml import html, etree
>  from multiprocessing import Process, Queue, Value
>  from urllib.parse import urljoin, urlparse, parse_qs, quote_plus
> 
> -
>  PATH = os.path.dirname(os.path.realpath(__file__))
>  COOKIES_FILE = os.path.join(PATH, "cookies.json")
> 
> @@ -880,7 +879,9 @@ class SafariBooks:
>                  self._thread_download_css(css_url)
> 
>          else:
> -            self._start_multiprocessing(self._thread_download_css, self.css)
> +            for css_url in self.css:
> +                self._thread_download_css(css_url)
> +            #self._start_multiprocessing(self._thread_download_css, self.css)
> 
>      def collect_images(self):
>          if self.display.book_ad_info == 2:
> @@ -897,7 +898,9 @@ class SafariBooks:
>                  self._thread_download_images(image_url)
> 
>          else:
> -            self._start_multiprocessing(self._thread_download_images, self.images)
> +            #self._start_multiprocessing(self._thread_download_images, self.images)
> +            for image_url in self.images:
> +                self._thread_download_images(image_url)
> 
>      def create_content_opf(self):
> ```

I tried this fix with book ISBN 9780135262047 but I ended up with a partial ebook downloaded (some contents missing), even if no errors were displayed in the output log.

**Comment by kingfetty** [comment]
FYI, it appears this error is only occuring on Linux installs.  I notice you're multithreading on Linux and on windows appear to be single threading.  Could this be an issue with SSL sessions stomping on each other?  Perhaps the answer is to single-thread the CSS downloads?

**Comment by phith0n** [comment]
The problem is that you're re-using the same connection for both processes.

Reference: <https://stackoverflow.com/questions/3724900/python-ssl-problem-with-multiprocessing>

**Comment by diegorodriguezv** [comment]
@nchiarappa 
> I tried this fix with book ISBN 9780135262047 but I ended up with a partial ebook downloaded (some contents missing), even if no errors were displayed in the output log.

I tried this fix and downloaded the book you mention correctly. It has 2504 pages and 1614 images.

**Comment by TheSnoozer** [comment]
Please get the latest version, this remaining css or image downloading issue should be fixed by d339d9d
